Add `teleport networking` subprocess for port/agent/x11 forwarding

Joerger opened the pull request:
Add a unified networking subprocess to handle port/agent/x11 forwarding requests. This subprocess is designed to run as the local user being connected to, including any modifications done to that user via the node's PAM stack.

Fixes https://github.com/gravitational/teleport/issues/17029

Note: this also fixes agent/port forwarding on new auto-provisioned users, which previously failed on the first attempt (or always with `insecure-drop`). This worked with X11 forwarding before with a more intricate solution, which has been replaced with the simpler "run-as-user" method.

Closes https://github.com/gravitational/teleport/issues/44479
Closes https://github.com/gravitational/teleport/issues/43623

Joerger committed:
Add networking subprocess for port and agent forwarding.

Joerger committed:
Remove extraneous x11 forwarding logic.

Joerger committed:
Fix lint.

Joerger committed:
Try create host user before starting networking process.

espadolini commented on lib/srv/reexec.go:
```diff
@@ -764,39 +607,208 @@ func runForward(handler forwardHandler) (errw io.Writer, code int, err error) {
+			var req networking.Request
+			if err := json.Unmarshal(buf[:n], &req); err != nil {
+				slog.With("error", err).ErrorContext(ctx, "Error parsing networking request.")
+				return
```
This leaks the files in `fbuf[:fn]`; it'd be cleaner if this closure was a function that took a payload and a file, or a payload and a pre-parsed control socket (the "outer" protocol with the json and socketpair doesn't depend on the request, after all).

espadolini commented on lib/srv/reexec.go:
```diff
@@ -764,39 +607,208 @@ func runForward(handler forwardHandler) (errw io.Writer, code int, err error) {
+			defer controlConn.Close()
+
+			log.Debug("Handling networking request")
+			netFile, err := handleNetworkingRequest(ctx, req)
```
This context should get closed by a read on `controlConn`:
```suggestion
			ctx, cancel := context.WithCancel(ctx)
			go func() {
				defer cancel()
				_, _ = controlConn.Read(make([]byte, 1))
			}()
			netFile, err := handleNetworkingRequest(ctx, req)
```

espadolini commented on lib/srv/reexec.go:
```diff
@@ -764,39 +607,208 @@ func runForward(handler forwardHandler) (errw io.Writer, code int, err error) {
+			return nil, trace.Wrap(err)
+		}
+
+		// Update the agent forwarding socket with more restrictive permissions.
+		if err = os.Chmod(socketPath, teleport.FileMaskOwnerOnly); err != nil {
```
If this is necessary, this is vulnerable to a connect() between listen() and chmod(); the directory containing the socket should be created with restrictive permissions instead.

Joerger replied:
You're right, this is how ssh does it too.

espadolini commented on lib/srv/reexec.go:
```diff
@@ -764,39 +607,208 @@ func runForward(handler forwardHandler) (errw io.Writer, code int, err error) {
+			var err error
+			err2 := conn.Control(func(descriptor uintptr) {
+				// Disable address reuse to prevent socket replacement.
+				err = syscall.SetsockoptInt(int(descriptor), syscall.SOL_SOCKET, syscall.SO_REUSEADDR, 0)
```
Is SO_REUSEADDR ever enabled by default?

Joerger replied:
Re-tested this and it seems it's only enabled by default for tcp listeners.

espadolini replied:
Is this necessary? If so, for what? `SO_REUSEADDR` allows listeners to be opened on top of sockets in `TIME_WAIT` state, opening multiple listeners on the same bind address at the same time is done with `SO_REUSEPORT`.

Joerger replied:
I'm not sure, I've preserved this from the remote port forwarding implementation. @atburke is this check necessary for some purpose or can it be removed?

atburke replied:
I believe `SO_REUSEADDR` was added here to prevent some malicious process from being able to hijack the address. You'd probably be able to get a more complete answer from @jentfoo.

espadolini commented on lib/srv/reexec.go:
```diff
@@ -764,39 +607,208 @@ func runForward(handler forwardHandler) (errw io.Writer, code int, err error) {
+		connFD, err := getConnFile(conn)
+		return connFD, trace.Wrap(err)
+
+	case networking.NetworkingOperationListen:
```
Should we add an operation for listenpacket too?

Joerger replied:
We don't have a use for it yet, do we?

espadolini reviewed (commented):
lib/sshutils/networking would benefit from some package-level docstring to explain the IPC protocol, since socketpairs passed around like this can be a bit confusing for the reader.

Joerger committed:
Run networking subprocess as root but change to user login after PAM call.

Joerger committed:
Resolve comments

Joerger commented on lib/sshutils/networking/validate_unix.go:
```diff
@@ -48,14 +46,14 @@ func validateListenerSocket(_ *srv.ServerContext, _ *net.UnixConn, listenerFD *o
+			return trace.Wrap(err)
 		} else if reusePort != 0 {
 			return trace.AccessDenied("SO_REUSEPORT is enabled on the socket")
 		}
 
 		// Verify that the listener is already listening.
 		if acceptConn, err := unix.GetsockoptInt(int(fd), unix.SOL_SOCKET, unix.SO_ACCEPTCONN); err != nil {
-			return trace.Wrap(err)
+			// Some systems may not support SO_ACCEPTCONN, so we ignore the error here
```
On my machine I am getting `protocol not available` for `SO_ACCEPTCONN`, and not for `SO_REUSEPORT`. I think I meant to make both options ignore the error.

Joerger committed:
Add x11 forwarding to the networking subprocess.

Joerger committed:
Fix networking process memory leak.

Joerger committed:
Update agent forwarding socket dir permissions.

Joerger committed:
Handle networking requests synchronously to maintain the current PAM context.

Joerger committed:
Use user environment in networking process.

espadolini commented on lib/srv/reexec.go:
```diff
@@ -764,39 +665,184 @@ func runForward(handler forwardHandler) (errw io.Writer, code int, err error) {
+			continue
+		}
+
+		// Starting a new goroutine takes us out of the current PAM context,
+		// so we handle requests synchronously.
+		handleNetworkingRequest(ctx, controlConn, buf[:n])
```
This will absolutely destroy the performance of ssh port forwarding.

What needs to happen in the pam thread? The creation of the socket? Or the actual connect/bind/listen?

Joerger replied:
> What needs to happen in the pam thread? The creation of the socket? Or the actual connect/bind/listen?

This is specifically meant to fix pam_namespace, so anything to do with the namespace. Currently that's just creating the temp files (`/tmp/.X11-unix` and /`tmp/teleport-xxx/socket-xxx`) and the `xauth` commands which write to `$HOME/.Xauthority`.

I managed to get a working fix by unlocking the OS thread [which we lock on init for PAM](https://github.com/gravitational/teleport/pull/4133).

Joerger committed:
Fix issue with PAM thread state being locked to the main goroutine; Clean up homeDir logic; Fix PAM thread state for xauth commands.

espadolini commented on lib/srv/reexec.go:
```diff
@@ -662,6 +657,11 @@ func RunNetworking() (errw io.Writer, code int, err error) {
 	ctx, cancel := context.WithCancel(context.Background())
 	defer cancel()
 
+	// Unlock OS thread (if locked by PAM) so that goroutines below inherit
+	// thread-state related to PAM, e.g. pam_namespace. See lib/pam/pam.go
+	// for details on why we lock the OS thread on init.
```
`man pam_namespace` explicitly says "The module must not be called from multithreaded processes." so I'm not really sure what it means to "inherit thread state related to PAM".

Joerger replied:
I'm not really sure how to understand it either, but sounds like this solution might have some hidden side effects since it's being explicitly warned against, thanks for catching that in the man page.

If we handle networking requests sequentially, the main downside is that dialing a TCP address with significant connection latency would prevent new requests, while tcp/unix listeners and unix dials should be quite fast. Should we just wrap the TCP dial handling in a goroutine alone? The only other solution I can think of would be to separate the subprocesses back out, or having the networking process create additional subprocesses when needed.

espadolini replied:
What does openssh do? That's the usual guideline when dealing with the low level system integration 🤣

Joerger replied:
I tried looking through the OpenSSH source code but I don't have enough C comprehension to understand how it managed to get around this multithreading issue. My current understanding is that the daemon creates a PAM auth thread for each user session, and any exec/forwarding requests within that session are ran with the PAM auth context. It's unclear to me how exactly that auth context is applied to calls to bind. Could the pam_namespace C library be wrapping it somehow? If you have more C experience I'd appreciate your insight.

WDYT about having the request handling loop run in a goroutine? Then we can have a bind-handling loop in the main thread which handles bind requests from the request handling goroutine? (Assuming bind is the only thing necessary from within the pam auth context)

Any other suggestions would be appreciated 😃

Joerger committed:
Selectively handle some networking requests in goroutines.

espadolini commented on lib/srv/reexec.go:
```diff
@@ -764,39 +660,203 @@ func runForward(handler forwardHandler) (errw io.Writer, code int, err error) {
+			default:
+				// Note: we don't currently support non-tcp network forwarding, so this branch is not
+				// currently reached. If in the future we add unix socket forwarding similar to OpenSSH's
+				// direct-streamlocal@openssh.com extension, we should revisit this multithreading limitation
+				// to prevent performance degradation.
```
for future readers: possible options are:
- doing manual async operations with nonblocking `connect` and `select`/`epoll`
- opening some amount of goroutines that lock the OS threa and open a pam context, and fanning out operations to be done in one of these PAM-enabled threads

espadolini commented on lib/srv/reexec.go:
```diff
@@ -764,39 +660,203 @@ func runForward(handler forwardHandler) (errw io.Writer, code int, err error) {
+		case networking.NetworkingOperationDial, networking.NetworkingOperationListen:
+			switch req.Network {
+			case "tcp":
+				// There are currently no known issues with tcp listen/dial in a multithreaded PAM context.
+				go handleNetworkingRequest(ctx, controlConn, req)
```
Until someone comes along and really wants to use a random pam module that does network namespacing 😬

Joerger commented on lib/srv/reexec.go:
```diff
@@ -740,19 +586,69 @@ func runForward(handler forwardHandler) (errw io.Writer, code int, err error) {
+		for i, g := range cred.Groups {
+			groups[i] = int(g)
+		}
+		if err := unix.Setgroups(groups); err != nil {
+			return errorWriter, teleport.RemoteCommandFailure, trace.Wrap(err, "failed to set groups for networking process")
+		}
+	}
+	if err := unix.Setgid(int(cred.Gid)); err != nil {
+		return errorWriter, teleport.RemoteCommandFailure, trace.Wrap(err, "failed to set gid for networking process")
+	}
+	if err := unix.Setuid(int(cred.Uid)); err != nil {
+		return errorWriter, teleport.RemoteCommandFailure, trace.Wrap(err, "failed to set uid for networking process")
```
I don't think so, as they shouldn't be I/O blocking operations.

Joerger committed:
Address comments.

espadolini commented on lib/sshutils/networking/networking.go:
```diff
@@ -0,0 +1,236 @@
+	unixListener.SetUnlinkOnClose(true)
+	return unixListener, nil
+}
+
+const requestBufferSize = 1024
```
```suggestion

// requestBufferSize is the maximum amount of data we're comfortable writing at
// once through a default unix datagram socket. Should fit comfortably in both
// linux and darwin with default settings.
const requestBufferSize = 1024
```

espadolini commented on lib/sshutils/networking/networking.go:
```diff
@@ -0,0 +1,236 @@
+	ctx, cancel := context.WithTimeout(ctx, 30*time.Second)
+	defer cancel()
+
+	slog.With("request", req).Debug("Sending networking request to child process")
```
+1

espadolini commented on lib/sshutils/networking/networking.go:
```diff
@@ -0,0 +1,236 @@
+		return nil, trace.Wrap(err)
+	}
+
+	listener, err := net.FileListener(listenerFD)
+	return listener, trace.Wrap(err)
```
```suggestion
	listener, err := net.FileListener(listenerFD)
	if err != nil {
		return nil, trace.Wrap(err)
	}
	
	if unixListener, ok := listener.(*net.UnixListener); ok {
		unixListener.SetUnlinkOnClose(true)
	}
	
	return listener, nil
```

espadolini commented on lib/sshutils/networking/networking.go:
```diff
@@ -0,0 +1,236 @@
+	})
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
```
```suggestion
	}
	defer listenerFD.Close()

```

espadolini commented on lib/sshutils/networking/networking.go:
```diff
@@ -0,0 +1,236 @@
+	})
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
```
```suggestion
	}
	defer connFD.Close()

```

espadolini commented on lib/sshutils/networking/networking.go:
```diff
@@ -0,0 +1,236 @@
+	})
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
```
```suggestion
	}
	defer listenerFD.Close()

```

espadolini commented on lib/sshutils/networking/networking.go:
```diff
@@ -0,0 +1,236 @@
+	})
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
```
```suggestion
	}
	defer listenerFD.Close()

```

espadolini commented on lib/srv/reexec.go:
```diff
@@ -761,42 +662,223 @@ func runForward(handler forwardHandler) (errw io.Writer, code int, err error) {
+		buf := make([]byte, 1)
+		// Wait for the terminate file descriptor to be closed. The FD will be closed when Teleport
+		// parent process wants to terminate the remote command and all childs.
+		_, err := termiantefd.Read(buf)
```
What's this (misspelled! 😬) pipe for? Can't a "graceful" shutdown just close the connection? (an ungraceful shutdown can just terminate the child I suppose)

espadolini reviewed (commented):
Still 100% unconvinced about the utility of unsetting `SO_REUSEADDR` or of `validateListenerSocket` in general.

Joerger committed:
Use an interrupt signal to trigger graceful exit in the networking subprocess.

Joerger committed:
Cleanup unix sockets from the child namespace.

Joerger committed:
Address comments.

Joerger committed:
Remove extraneous network forwarding socket validation.

Joerger commented:
> Still 100% unconvinced about the utility of unsetting `SO_REUSEADDR` or of `validateListenerSocket` in general.

I'd tend to agree, especially now that we always open listeners from the user's permissions. I've removed both in [f4b0339](https://github.com/gravitational/teleport/pull/43756/commits/f4b0339ddd70063a04f7a27e2fdb47a67db889ee).

Joerger committed:
Fix tests and lint.

Joerger committed:
Re-add networking process done channel to avoid deadlocks on reading unix sockets passed to a dead process.

Joerger committed:
Remove broken test.

espadolini commented on lib/srv/reexec.go:
```diff
@@ -761,42 +660,241 @@ func runForward(handler forwardHandler) (errw io.Writer, code int, err error) {
+				go func() {
+					cleanup := handleNetworkingRequest(ctx, controlConn, req)
+					if cleanup != nil {
+						context.AfterFunc(ctx, func() {
```
This (and the defer later) are 100% not going to work, they will accumulate memory for each connection for the lifetime of the process even for short lived connections.

If we need to keep a list of paths to be deleted it should be an explicit list of paths and we should probably have some limit around how many we end up dealing with.

Joerger replied:
This cleanup is only for agent/x11/remote forwarding _unix_ listeners, not local forwarding connections or remote forwarding tcp listeners. IMO the only thing missing here might be handling "cancel-tcpip-forward" for _unix listeners which we don't support_. As it currently stands each networking process will defer at most 2 cleanups, one for x11 and one for agent forwarding. 

I do agree that handling this more explicitly is a good idea, with some stored list of sockets to close, I just am not convinced of its necessity at this time.

espadolini replied:
The main issue is that the cleanup will run at the end of the networking process, not when the listener is closed. And wasn't there an issue we recently fixed because we assumed that there was only going to be one X11 forward request in the lifetime of a connection and there most definitely wasn't?

If we are always supposed to unlink the listener when it's closed, maybe we can just keep the per-request socket open on both sides, and close it after closing the listener, to signal the child to close it?

A really cursed idea for the future: when we pass back the unix listener, we also pass a file descriptor for the directory that contains it, and then no matter the namespace we can `unlinkat` the socket after closing it, from the parent process 👀

espadolini commented on lib/srv/reexec.go:
```diff
@@ -761,42 +660,241 @@ func runForward(handler forwardHandler) (errw io.Writer, code int, err error) {
+
+	case networking.NetworkingOperationListenAgent:
+		// Create a temp directory to hold the agent socket.
+		sockDir, err := os.MkdirTemp(os.TempDir(), "teleport-")
```
```suggestion
		sockDir, err := os.MkdirTemp("", "teleport-")
```

Joerger committed:
Address comments.

Joerger commented on lib/srv/reexec.go:
```diff
@@ -761,42 +660,241 @@ func runForward(handler forwardHandler) (errw io.Writer, code int, err error) {
+			return nil, nil, trace.Wrap(err)
+		}
+
+		return listenerFD, cleanupDir, nil
```
`os.RemoveAll(sockDir)` should be enough, right?

espadolini replied:
I guess? If that's what we're doing right now it's fine to keep it that way.

Joerger committed:
Cleanup; remove unused child error file.

espadolini commented on lib/sshutils/networking/networking.go:
```diff
@@ -0,0 +1,399 @@
+	defer remoteFD.Close()
+	cmd.ExtraFiles = append(cmd.ExtraFiles, remoteFD)
+
+	// Propagate stderr from the spawned Teleport process to log any errors.
```
If the child has stderr from the parent to log errors, why do we need the parent to listen on the control connection for errors?

Joerger replied:
Logging in the child process just wouldn't working, I tried linking stdout in addition to stderr but no logs showed up. I also tested the logs in the exec process never show up. I couldn't figure out the root issue in a reasonable amount of time so I made this change and moved on.

espadolini commented on lib/sshutils/networking/networking.go:
```diff
@@ -0,0 +1,399 @@
+	default:
+		// We use an interrupt signal because the main control socket is not connection oriented,
+		// meaning reading from the child side won't unblock with the parent closes it.
+		p.cmd.Process.Signal(os.Interrupt)
```
This is all fine and good until the parent dies without sending the signal - one of the good properties of the pipe trick (which we are 100% screwing up on the parent side by keeping both sides of the pipes in `srv.ServerContext` alive) is that when the other side that's holding the pipe dies for any reason, the pipe unblocks.

espadolini commented on lib/sshutils/networking/networking.go:
```diff
@@ -0,0 +1,399 @@
+}
+
+// childConn is a net.Conn backed by a connection in the child process.
+type childConn struct {
```
Wrappers have a cost, especially when dealing with OS resources that the stdlib often type asserts on (to optimize copies into splices, for example). Doubling the amount of file descriptors for a fairly active "bastion host" that does mostly networking also has a not insignificant cost.

I'm not sure if there's a great way other than having both sides just know about the correct behavior on a per-request basis, but I would keep this cleanup-on-request-connection-drop construction only for the listeners that need it (unix listeners - which are currently not even used - and the X11 and agent listeners).

espadolini commented on lib/sshutils/networking/networking.go:
```diff
@@ -0,0 +1,399 @@
+	}
+	remoteFD.Close()
+
+	// Read in another goroutine so we can cancel it if the networking process stops or the context closes.
```
Why? Just tie the context to the closing of the connection with a `context.AfterFunc` (which returns a function you can call to undo the tying).

Joerger replied:
We don't want to close the per-request connection if the request succeeds, as the networking resources's lifetime is tied to the connection. We could add a conditional close if that's better? Seems iffy.

```go
func (p *Process) sendRequest(ctx context.Context, req Request) (f *childNetworkingFile, err error) {
    context.AfterFunc(func(){
        if err != nil {
            reqConn.Close()
        )
    })
```

espadolini replied:
You can cancel a context AfterFunc, which we should take advantage of.

espadolini commented on lib/sshutils/networking/networking.go:
```diff
@@ -0,0 +1,399 @@
+		defer close(fileCh)
+		file, err := readResponse(requestConn)
+		if err != nil {
+			errC <- err
+		} else {
+			fileCh <- file
```
Case in point to avoid goroutines if not strictly necessary: this will block forever (leaking the goroutine stack and the entry in the scheduler) if an object or an error is being pushed into the channels and the context or p.done unblock, so nothing is reading in any of those two channels anymore.

espadolini commented on lib/sshutils/networking/networking.go:
```diff
@@ -0,0 +1,399 @@
+	fileCh := make(chan *os.File)
+	errC := make(chan error)
+	go func() {
+		defer close(fileCh)
```
Why this close? Why only on `fileCh`?

espadolini commented on lib/srv/reexec.go:
```diff
@@ -740,63 +573,380 @@ func runForward(handler forwardHandler) (errw io.Writer, code int, err error) {
+		}
+	}()
+
+	// Some requests and cleanup must happen in the main PAM thread. Handle any requests
+	// or errors passed from the main request goroutine above, post connection cleanup,
+	// or interrupt signals from the parent process.
```
I would make this a bit more generic and have the goroutine (tied to the PAM thread) range over a `chan func()` that's "owned" by the side goroutine and execute whatever comes through, then return whatever's in some variables (as set by the side goroutine before closing the channel).

espadolini commented on lib/srv/reexec.go:
```diff
@@ -740,63 +573,380 @@ func runForward(handler forwardHandler) (errw io.Writer, code int, err error) {
+	cleanupC := make(chan networkingResource)
+	defer func() {
+		go func() {
+			wg.Wait()
```
I don't believe the sequencing is strong enough that this can't race with a `wg.Add` (which, at 0, would cause a panic). For example, `handleNetworkingRequest` for a TCP dial can return a success and then the context can get cancelled by a signal, causing the loop to exit and the defer to run, while the code after `handleNetworkingRequest` then calls `wg.Add`.

espadolini replied:
(I'm not saying we should fix this code to handle that specific sequence, I'm saying that this code should be reworked so this entire class of confusing concurrency problems goes away)

espadolini commented on lib/srv/reexec.go:
```diff
@@ -740,63 +573,380 @@ func runForward(handler forwardHandler) (errw io.Writer, code int, err error) {
+				writeErrorToConn(parentConn, trace.BadParameter("failed to get a connection from control file"))
+				continue
+			}
+			context.AfterFunc(ctx, func() { req.conn.Close() })
```
This should be undone once we're done with `req.conn`, or it'll accumulate callbacks in the context forever.

espadolini commented:
Since the agent listener is in a temp path and, from what I understand, the X11 listener can overstay its welcome without too many issues (other listeners will just take bigger numbers), what if we punt on "instant" cleanup and leave the cleanup of on-disk sockets for the end of the process? We're likely going to have to revisit this if we add the ability to listen on unix sockets, but it's probably fine for now.

I'd still keep an explicit list of paths to delete rather than do it in the defer stack, and leave the closing of the listener to the parent process tho.

Joerger committed:
- Explicility list file paths to clean up at the end of the networking process.
- Use terminate FD instead of interrupt to signal the process to exit.
- Log errors from the parent process.
- Read networking request response without goroutine.

espadolini commented on lib/srv/reexec.go:
```diff
@@ -740,63 +573,310 @@ func runForward(handler forwardHandler) (errw io.Writer, code int, err error) {
+			writeErrorToConn(parentConn, trace.Wrap(err, "failed to get a connection from control file"))
+			continue
+		}
+		context.AfterFunc(ctx, func() { requestConn.Close() })
```
This will accumulate in the `context.cancelCtx` for every new request. The returned cleanup function must be called when we're done with the `requestConn` if this is used in a loop.

espadolini commented on lib/srv/reexec.go:
```diff
@@ -740,63 +573,310 @@ func runForward(handler forwardHandler) (errw io.Writer, code int, err error) {
+			return os.RemoveAll(sockDir)
+		}
+
+		// Update the agent forwarding socket dir with more restrictive permissions.
+		if err = os.Chmod(sockDir, teleport.PrivateDirMode); err != nil {
```
I think that MkdirTemp already creates the directory with 0o700 perms.

espadolini commented on lib/sshutils/networking/networking.go:
```diff
@@ -0,0 +1,341 @@
+	go func() {
+		for {
+			buf := make([]byte, RequestBufferSize)
+			n, err := p.conn.Read(buf)
```
This is a stream, you don't know where the boundaries of the message are. Why is this a stream?

Joerger replied:
I changed it to a stream so that closing it would be reflected in the child process, rather than managing an extra fd for closure. I can revert https://github.com/gravitational/teleport/commit/dc76a7bb3f22f49baef45d59775eb0b7c4614405

Joerger replied:
I see the issue with this now, reverting. Thanks.

espadolini commented on lib/sshutils/networking/networking.go:
```diff
@@ -0,0 +1,341 @@
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	context.AfterFunc(ctx, func() { requestConn.Close() })
```
```suggestion
	defer context.AfterFunc(ctx, func() { requestConn.Close() })()
```

espadolini replied:
Except that now the request connection is datagram so closing it is no longer a signal for the child process, so the context in any of the operations here is a lie.

Joerger replied:
I changed it to a stream so that closing it would be reflected in the child process, rather than managing an extra fd for closure. I can revert https://github.com/gravitational/teleport/pull/43756/commits/dc76a7bb3f22f49baef45d59775eb0b7c4614405

Joerger replied:
Reverted to stream socket. Why would this need to be deferred? we want to unblock read on the request conn if the context is closed.

Joerger replied:
I ended up having to remove this connection-context logic anyways due to it causing an x11 test to be flaky - [42095d6](https://github.com/gravitational/teleport/pull/43756/commits/42095d6e49cbaca8ed4c1595ba61848d11cc25a7). I couldn't figure out what the root cause of the flake was, some test environment fishiness probably. Once that's sorted we could revert [42095d6](https://github.com/gravitational/teleport/pull/43756/commits/42095d6e49cbaca8ed4c1595ba61848d11cc25a7) back.

Joerger replied:
Nevermind I found the culprit b9062c73e0dda3793eecf94b722007b7cade9598

espadolini reviewed (commented):
Why is the control connection a stream now? Why are the request connections datagram?

Joerger committed:
Cleanup.

Joerger committed:
Make request socket a stream.

Joerger committed:
Remove extraneious Chmod.

espadolini commented on lib/sshutils/networking/networking.go:
```diff
@@ -0,0 +1,335 @@
+		return nil, trace.BadParameter("expected *net.UnixListener but got %T", listener)
+	}
+
+	unixListener.SetUnlinkOnClose(true)
```
We have to leave the unlink to the child process' graceful termination, right?
```suggestion
```

Joerger replied:
If the parent process can unlink the listener it should and can - the child will still try to cleanup the socket files if the unlink fails due to pam_namespace

espadolini replied:
The parent process (running as root, most likely) should not be unlinking paths that might belong to other sockets tho, and when fs namespaces are in the mix, there's no telling what the path local to the namespace actually is from the parent's point of view - and there's no telling if something else is there instead!

Joerger replied:
Good point...

espadolini commented on lib/sshutils/networking/networking.go:
```diff
@@ -0,0 +1,335 @@
+		return nil, trace.Wrap(err)
+	}
+
+	if unixListener, ok := listener.(*net.UnixListener); ok {
+		unixListener.SetUnlinkOnClose(true)
+	}
+
```
```suggestion
```

espadolini commented on lib/sshutils/networking/networking.go:
```diff
@@ -0,0 +1,335 @@
+		return nil, trace.BadParameter("expected *net.UnixListener but got %T", listener)
+	}
+
+	unixListener.SetUnlinkOnClose(true)
```
```suggestion
```

espadolini commented on lib/srv/reexec.go:
```diff
@@ -740,63 +578,296 @@ func runForward(handler forwardHandler) (errw io.Writer, code int, err error) {
+
+		var req networking.Request
+		if err := json.Unmarshal(buf[:n], &req); err != nil {
+			writeErrorToConn(parentConn, trace.Wrap(err, "error parsing networking request"))
```
This can be an error returned by the request I think?
```suggestion
			writeErrorToConn(requestConn, trace.Wrap(err, "error parsing networking request"))
```

Joerger replied:
But the requestConn won't respect the error message boundaries since it's a stream, right?

espadolini replied:
That's true - but we can just read the whole connection until the EOF after we try to read the file descriptor; if we don't do that we're kinda just relying on luck to get decent error messages for the user (the message is propagated into the response to the direct-tcpip request, right?)

espadolini reviewed (approved):
I think this is correct save for the last few minor comments, but someone else should probably also take another good look at this.

Joerger committed:
Send request level error to request conn.

Joerger committed:
Don't unlink unix sockets from the parent process.

Joerger committed:
Read full error message from stream.

rosstimothy reviewed (commented):
I noticed that error messages are now different if the user does not have access to the target host.

Before:
```bash

$ ./releases/15.4.4/tsh ssh -R 3303:127.0.01:3000 jim@server01
ERROR: access denied to jim connecting to server01:0

$ ./releases/15.4.4/tsh ssh -D 8080 jim@server01
ERROR: access denied to jim connecting to server01:0

$ ./releases/15.4.4/tsh ssh -L 3303:127.0.01:3000 jim@server01
ERROR: access denied to jim connecting to server01:0
```

This PR:
```bash

$ ./releases/15.4.4/tsh ssh -R 3303:127.0.01:3000 jim@server01
ERROR: error reading from server: EOF

$ ./releases/15.4.4/tsh ssh -D 8080 jim@server01
ERROR: error reading from server: EOF

$ ./releases/15.4.4/tsh ssh -L 3303:127.0.01:3000 jim@server01
ERROR: error reading from server: EOF

```

espadolini commented on lib/sshutils/networking/networking.go:
```diff
@@ -0,0 +1,318 @@
+		if n > 0 {
+			// The networking process only ever writes to the request conn if an error occurs.
+			// Read the rest of the connection to ensure we don't return just a partial stream.
+			errMsg, err := io.ReadAll(conn)
```
😅

```suggestion
			errMsg, err := io.ReadAll(io.LimitReader(conn, cap(buf) - len(buf)))
```

Joerger committed:
Remove remaining logs in child process.

Joerger commented:
> I noticed that error messages are now different if the user does not have access to the target host.
> 
> Before:
> 
> ```shell
> $ ./releases/15.4.4/tsh ssh -R 3303:127.0.01:3000 jim@server01
> ERROR: access denied to jim connecting to server01:0
> 
> $ ./releases/15.4.4/tsh ssh -D 8080 jim@server01
> ERROR: access denied to jim connecting to server01:0
> 
> $ ./releases/15.4.4/tsh ssh -L 3303:127.0.01:3000 jim@server01
> ERROR: access denied to jim connecting to server01:0
> ```
> 
> This PR:
> 
> ```shell
> $ ./releases/15.4.4/tsh ssh -R 3303:127.0.01:3000 jim@server01
> ERROR: error reading from server: EOF
> 
> $ ./releases/15.4.4/tsh ssh -D 8080 jim@server01
> ERROR: error reading from server: EOF
> 
> $ ./releases/15.4.4/tsh ssh -L 3303:127.0.01:3000 jim@server01
> ERROR: error reading from server: EOF
> ```



> I noticed that error messages are now different if the user does not have access to the target host.
> 
> Before:
> 
> ```shell
> $ ./releases/15.4.4/tsh ssh -R 3303:127.0.01:3000 jim@server01
> ERROR: access denied to jim connecting to server01:0
> 
> $ ./releases/15.4.4/tsh ssh -D 8080 jim@server01
> ERROR: access denied to jim connecting to server01:0
> 
> $ ./releases/15.4.4/tsh ssh -L 3303:127.0.01:3000 jim@server01
> ERROR: access denied to jim connecting to server01:0
> ```
> 
> This PR:
> 
> ```shell
> $ ./releases/15.4.4/tsh ssh -R 3303:127.0.01:3000 jim@server01
> ERROR: error reading from server: EOF
> 
> $ ./releases/15.4.4/tsh ssh -D 8080 jim@server01
> ERROR: error reading from server: EOF
> 
> $ ./releases/15.4.4/tsh ssh -L 3303:127.0.01:3000 jim@server01
> ERROR: error reading from server: EOF
> ```

I'm seeing this on master too, I don't think it's related to this PR.

Joerger committed:
Merge branch 'master' into joerger/teleport-networking-subprocess

atburke reviewed (approved):
LGTM, just a few typos.

flyinghermit committed:
web: support SAML resource deletion in unified resources view (#44311)

* update SAML app edit mechanism to accomodate both edit and delete functions

* use user saml idp access to disable menu actions

* add comment to ResourceActionButton props

* refactor: SamlAppActionContext to edit and delete Saml application

* address review comments

Joerger committed:
Fix typos.

Joerger committed:
Add networking process tests to replace old tests.

Joerger committed:
Don't close remote file descriptor before the child process has a chance to connect to it.

Joerger committed:
Fix lint; skip broken test.

Joerger committed:
Disable broken test and restore older test to cover for it.

Joerger committed:
Remove unused agent forwarding test.

Joerger committed:
Merge branch 'master' into joerger/teleport-networking-subprocess

Joerger committed:
Fix lint, replace broken test.

Joerger committed:
Fix old test that used testify/require in goroutine.

Joerger committed:
Close request context immediately to avoid deadlock (extended timeout) on reading from the closed process.

Joerger committed:
Merge branch 'master' into joerger/teleport-networking-subprocess

Joerger commented:
@rosstimothy All tests are passing, just needs a flaky test skip. I double checked that `TestRootNetworkingCommand` isn't flaky, it just ends with an error when the flaky test detector times out at 10m.

Joerger committed:
Merge branch 'master' into joerger/teleport-networking-subprocess
